### PR TITLE
Implemented kelp

### PR DESCRIPTION
**Pull request by ShockedPlot7560**
## Introduction
Implemented Kelp

### API changes
* Add kelp in registry
### Behavioural changes
N/A

## Tests
[https://youtu.be/LbO1iiyFZX8](https://youtu.be/LbO1iiyFZX8)

### CODE REVIEW

File: src/item/VanillaItems.php [hunk 1]
```diff
@@ -133,6 +133,7 @@
  * @method static Item DRAGON_BREATH()
  * @method static Skull DRAGON_HEAD()
  * @method static DriedKelp DRIED_KELP()
+ * @method static ItemBlock KELP()
```
Comment (dktapps) [inline src/item/VanillaItems.php]: We don't put blocks in VanillaItems. `VanillaBlocks::KELP()->asItem()` would be used instead.

File: src/item/StringToItemParser.php [hunk 2]
```diff
@@ -615,6 +615,7 @@ private static function make() : self{
 		$result->registerBlock("jungle_wood", fn() => VanillaBlocks::JUNGLE_WOOD());
 		$result->registerBlock("jungle_wood_stairs", fn() => VanillaBlocks::JUNGLE_STAIRS());
 		$result->registerBlock("jungle_wooden_stairs", fn() => VanillaBlocks::JUNGLE_STAIRS());
+		$result->registerBlock("kelp_block", fn() => VanillaBlocks::KELP());
```
Comment (dktapps) [inline src/item/StringToItemParser.php]: StringToItemParser should not include blocks that can't be obtained in item form.

File: src/block/Kelp.php [hunk 3]
```diff
@@ -128,4 +130,11 @@ private function grow(Kelp $block) : bool{
 		}
 		return false;
 	}
+
+	public function asItem(): Item{
+		return ItemFactory::getInstance()->get(
+			ItemIds::KELP,
```
Comment (dktapps) [inline src/block/Kelp.php]: This should be specified in the `BlockIdentifier` constructor. It shouldn't be needed here.

Comment (ShockedPlot7560) [reply src/block/Kelp.php]: Indeed, oops

File: src/block/VanillaBlocks.php [hunk 4]
```diff
@@ -163,6 +163,7 @@
  * @method static DoubleTallGrass DOUBLE_TALLGRASS()
  * @method static DragonEgg DRAGON_EGG()
  * @method static DriedKelp DRIED_KELP()
+ * @method static Kelp KELP()
```
Comment (JavierLeon9966) [inline src/block/VanillaBlocks.php]: This should be below `JUNGLE_WOOD()`

File: src/item/ItemFactory.php [hunk 5]
```diff
@@ -96,6 +96,7 @@ public function __construct(){
 		$this->register(new CookedSalmon(new ItemIdentifier(ItemIds::COOKED_SALMON, 0), "Cooked Salmon"));
 		$this->register(new Cookie(new ItemIdentifier(ItemIds::COOKIE, 0), "Cookie"));
 		$this->register(new DriedKelp(new ItemIdentifier(ItemIds::DRIED_KELP, 0), "Dried Kelp"));
+		$this->register(new ItemBlock(new ItemIdentifier(ItemIds::KELP, 0), VanillaBlocks::KELP()));
```
Comment (JavierLeon9966) [inline src/item/ItemFactory.php]: This should be below `ItemIds::IRON_NUGGET`

File: src/item/StringToItemParser.php [hunk 6]
```diff
@@ -1124,6 +1124,7 @@ private static function make() : self{
 		$result->register("iron_shovel", fn() => VanillaItems::IRON_SHOVEL());
 		$result->register("iron_sword", fn() => VanillaItems::IRON_SWORD());
 		$result->register("jungle_boat", fn() => VanillaItems::JUNGLE_BOAT());
+		$result->register("kelp", fn() => VanillaBlocks::KELP()->asItem());
```
Comment (JavierLeon9966) [inline src/item/StringToItemParser.php]:
```suggestion
		$result->registerBlock("kelp", fn() => VanillaBlocks::KELP());
```
And also it should be below this:
```php
$result->registerBlock("jungle_wooden_stairs", fn() => VanillaBlocks::JUNGLE_STAIRS());
```

**Review by JavierLeon9966 — CHANGES_REQUESTED** [review]

File: src/block/VanillaBlocks.php (hunk 4 above)
Comment (ShockedPlot7560) [reply src/block/VanillaBlocks.php]: These comments are not sorted alphabetically, are they?

Comment (JavierLeon9966) [reply src/block/VanillaBlocks.php]: They are as you can see: `DOUBLE_TALLGRASS()`, `DRAGON_EGG()`, `DRIED_KELP()`, etc.

Comment (dktapps) [reply src/block/VanillaBlocks.php]: You can generate the comment using `build/generate-registry-annotations.php`.

**Comment by ipad54** [comment]
After rejoining happens this, and you can restore your breathing in kelp, because PM not support waterlogging
Screenshot_20220107_134742_com mojang minecraftpe
Screenshot_20220107_134738_com mojang minecraftpe

**Comment by ShockedPlot7560** [comment]
True but its like sea pickles or coral, when waterlogged will be available, it will works normally.

**Comment by ipad54** [comment]
In vanilla kelps breaks if there isn't water nearby

File: src/block/Kelp.php [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+use pocketmine\math\Vector3;
+use pocketmine\player\Player;
+use pocketmine\world\BlockTransaction;
+use function mt_rand;
+
+class Kelp extends Transparent{
+
+	protected int $age = 0;
+
+	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
```
Comment (dktapps) [inline src/block/Kelp.php]: This should be further down with the rest of the on* functions.

**Review by dktapps — COMMENTED** [review]
This implementation is not currently compatible with PM due to the limited size of metadata used. 25 age states require 5 bits to store; PM's metadata storage only allows 4.

**Comment by dktapps** [comment]
Please retarget this to `next-major`

**Pull request closed without merging**